olp-cpp-sdk-core/tests/client/HRNTest.cpp

```diff
@@ -0,0 +1,151 @@
+      HRN hrn("hrn:here:schema:::group_id:artifact_id:version");
+      EXPECT_FALSE(hrn.IsNull());
+    }
+    {
+      SCOPED_TRACE("Invalid Schema HRN (missing group_id)");
+      HRN hrn("hrn:here:schema::test::artifact_id:version");
+      EXPECT_TRUE(hrn.IsNull());
+    }
+    {
+      SCOPED_TRACE("Invalid Schema HRN (missing artifact_id)");
+      HRN hrn("hrn:here:schema::test:group_id::version");
+      EXPECT_TRUE(hrn.IsNull());
+    }
+    {
+      SCOPED_TRACE("Invalid Schema HRN (missing version)");
+      HRN hrn("hrn:here:schema::test:group_id:artifact_id:");
+      EXPECT_TRUE(hrn.IsNull());
+    }
+  }
+  {
+    SCOPED_TRACE("Pipeline HRN Tests");
+    {
+      SCOPED_TRACE("Valid Pipeline HRN");
+      HRN hrn("hrn:here:pipeline::test:test_pipeline");
+      EXPECT_FALSE(hrn.IsNull());
+    }
+    {
+      SCOPED_TRACE("Valid Pipeline HRN without realm");
+      HRN hrn("hrn:here:pipeline:::test_pipeline");
+      EXPECT_FALSE(hrn.IsNull());
+    }
+    {
+      SCOPED_TRACE("Invalid Pipeline HRN (missing pipeline id)");
+      HRN hrn("hrn:here:pipeline::test:");
+      EXPECT_TRUE(hrn.IsNull());
+    }
+  }
+}
+
+TEST(HRNTest, CompareHRNs) {
+  SCOPED_TRACE("Compare equal HRNs");
+  {
+    SCOPED_TRACE("Compare equal Catalog HRNs");
+    EXPECT_EQ(HRN("hrn:here:data:::hereos-internal-test-v2"),
+              HRN("hrn:here:data:::hereos-internal-test-v2"));
+    EXPECT_NE(HRN("hrn:here:data:::hereos-internal-test-v1"),
+              HRN("hrn:here:data:::hereos-internal-test-v2"));
+    EXPECT_NE(HRN("hrn:here:data:::hereos-internal-test-v2"),
+              HRN("hrn:here:data::test:hereos-internal-test-v2"));
+  }
+  {
+    SCOPED_TRACE("Compare equal Schema HRNs");
+    EXPECT_EQ(HRN("hrn:here:schema:::group_id:artifact_id:version"),
+              HRN("hrn:here:schema:::group_id:artifact_id:version"));
+    EXPECT_NE(HRN("hrn:here:schema:::group_id:artifact_id:version_1"),
+              HRN("hrn:here:schema:::group_id:artifact_id:version_2"));
+    EXPECT_NE(HRN("hrn:here:schema:::group_id:artifact_id:version_1"),
+              HRN("hrn:here:schema::test:group_id:artifact_id:version_2"));
+  }
+  {
+    SCOPED_TRACE("Compare equal Pipeline HRNs");
+    EXPECT_EQ(HRN("hrn:here:data:::test_pipeline"),
+              HRN("hrn:here:data:::test_pipeline"));
+    EXPECT_NE(HRN("hrn:here:data:::test_pipeline_1"),
+              HRN("hrn:here:data:::test_pipeline_2"));
+    EXPECT_NE(HRN("hrn:here:data:::test_pipeline"),
+              HRN("hrn:here:data::test:test_pipeline"));
+  }
+}
+
+TEST(HRNTest, ToString) {
+  EXPECT_EQ(HRN("hrn:here:data:::hereos-internal-test-v2").ToString(),
+            "hrn:here:data:::hereos-internal-test-v2");
+  EXPECT_EQ(HRN("hrn:here:schema:::group_id:artifact_id:version").ToString(),
+            "hrn:here:schema:::group_id:artifact_id:version");
+  EXPECT_EQ(HRN("hrn:here:data:::test_pipeline").ToString(),
+            "hrn:here:data:::test_pipeline");
+}
+
```

Comment:
I miss tests for parsing. Means creating a HRN and then test that each element, e.g. partition, region, account, etc., is parsed correctly.